### ISSUE TITLE
Add -Wenum-conversion that works also on non-constexpr contexts

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#59045]
Clang recently added the error `-Wenum-constexpr-conversion` (downgradable to warning). This detects cases where casting an integer outside the valid range of an enum into said enum is UB. The compiler is required to diagnose this in constant expressions, where UB is not allowed.

I feel this is a quite useful warning that would also be good to have in regular, non-constexpr code (in that case, classified as an opt-in warning). Since the majority of the implementation for detecting this is already in place, it seems to me it should be fairly doable to add such a warning for the general case.

Example:

```cpp
enum Foo
{
  A=0,
  B,
};

constexpr Foo x = static_cast<Foo>(123);  // Hard error, UB (-Wenum-constexpr-conversion)

int main()
{
   Foo y = static_cast<Foo>(123);  // Warning (-Wenum-conversion)
}
```

What do you think, would this be feasible?

Thanks!
